### PR TITLE
Pull allocation logic out of executor

### PR DESCRIPTION
**Pull request by csarofeen**
The fusion executor file is littered with logic around allocations. I pulled that logic into a separate file. this PR is just code motion.

### CODE REVIEW

File: csrc/fusion_executor/allocations.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+std::vector<at::Tensor> allocateOutputs(
+    const Fusion* fusion,
+    const std::vector<GlobalBufferInfo>& output_info,
+    const c10::Device& device,
+    ExpressionEvaluator& ee);
+
+//! Used in distributed setting where we only want to
+//!  allocate output space and receive output data from
+//!  a different rank instead of computing them.
+std::vector<at::Tensor> allocOutputSpace(
```
Comment (csarofeen) [inline csrc/fusion_executor/allocations.h:83]: Duplicate declaration, need to remove.

**Comment by csarofeen** [comment]
!build

File: csrc/fusion_executor/allocations.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+};
+
+//! This function is useful for parallel compilation of segmented fusions.
+//! It returns non-allocated KernelArgumentHolder, representing the output
+//! sizes from kernel execution.
+//! Notes: 1. This API should ignore aliased outputs instead of
+//! pushing scalar int 0 as a place-holder.
+//! 2. This API does not allocate output in memory, but only returns the
+//! inferred output sizes. Used in kernel_cache.cpp.
+KernelArgumentHolder inferOutputSizes(
```
Comment (jjsjann123) [inline csrc/fusion_executor/allocations.h:35]: :clap:

**Review by jjsjann123 — APPROVED** [review]
LGTM.

**Pull request merged**